models.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+require 'json'
+require 'cgi'
+
+def parse_all_data
+  original_data = File.read('db.json')
+  data = JSON.parse(original_data, symbolize_names: true)
+  data[:memos] || []
+end
+
+def find_memo_data(id)
+  all_data = parse_all_data
+  all_data.find { |memo| memo[:id] == id }
+end
+
+def delete_memo_data(id)
+  old_data = parse_all_data
+  new_data = { 'memos' => old_data.reject { |memo| memo[:id] == id } }
+  File.open('db.json', 'w') do |file|
+    file.write(JSON.pretty_generate(new_data))
+  end
+end
+
+def edit_memo_data(params)
+  symbolized_params = params.transform_keys(&:to_sym)
+  all_data = parse_all_data
+  all_data.each do |memo|
+    if memo[:id] == symbolized_params[:id].to_i
+      memo[:name] = symbolized_params[:name]
+      memo[:text] = symbolized_params[:text]
+    end
+  end
+  new_data = { memos: all_data }
+  File.open('db.json', 'w') { |file| file.write(JSON.pretty_generate(new_data)) }
+end
+
+def set_id
+  all_data = parse_all_data
+  last_id = if all_data.empty?
+              0
+            else
+              all_data[-1][:id]
+            end
+  last_id + 1
+end
+
+def add_new_memo(params)
+  symbolized_params = params.transform_keys(&:to_sym)
+  id = set_id
+  all_data = parse_all_data
+  all_data.push({ name: symbolized_params[:name], text: symbolized_params[:text], id: })
```

Comment:
```suggestion
  all_data.push(symbolized_params.slice(:name, :text).merge(id:))
```

高等テクニックですがこんな書き方でも動きそうです。

Reply:
`merge`メソッドでキーと値を統合しながら、`push`の引数として使うことができるんですね！変更しました。https://github.com/SuzukaHori/sinatra_practice/pull/1/commits/bdd187f196f50c726dcaa9b640543ae51c9eb4fd

一点疑問に思ったのですが、一見`all_data.push(symbolized_params.merge(id:))`だけでも動きそうに見えました。`slice(:name, :text)`を入れているのにはどのような理由がありますか？

Reply:
たしかにsliceはなくてもほとんど問題なさそうですね。
特にこの課題では`symbolized_params.merge(id:)`でも良いと思います。

ただし、業務では`symbolized_params.merge(id:)`のようなコードを書くのはちょっと危険です。
というのも、`params`には悪意あるユーザーが好きなデータを送り込めてしまうからです。
つまり、リクエストに細工をすればnameやtext以外のキーと値をJSONに保存できてしまう、ということです。

じゃあ、そこからどんな悪用が考えられますか？と言われるとぱっと答えは出てこないのですが、JSONを開いたら見覚えのないキーと値が保存されているというのは混乱の元になりますし、無用なリスクを放置していることになります。

なので、paramsの値を盲目的にJSONに保存するのではなく、必要なキーだけに絞って保存することには意味があります。
ちなみに、Railsのstrong parametersという機能もほぼ同じ目的で存在しています。

Reply:
`slice`メソッドで保存されるキーが限定されることが、セキリュティ対策につながるということですね。確かに、入力されたparamsの値をそのまま保存するのは危険でした。
Railsはまだ序盤なのですが、strong parametersは見たことがあるような気がするので、確認します🙏